src/algorithms/libs/beidou_b1c_signal_processing.cc

```diff
@@ -0,0 +1,859 @@
+
+
+void beidou_b1c_pilot_sinboc_11_gen_int(gsl::span<int> _dest, gsl::span<const int> _prn)
+{
+    const uint32_t _length_in = BEIDOU_B1Cp_CODE_LENGTH_CHIPS;
+    auto _period = static_cast<uint32_t>(_dest.size() / _length_in);
+    for (uint32_t i = 0; i < _length_in; i++)
+        {
+            for (uint32_t j = 0; j < (_period / 2); j++)
+                {
+                    _dest[i * _period + j] = _prn[i];
+                }
+            for (uint32_t j = (_period / 2); j < _period; j++)
+                {
+                    _dest[i * _period + j] = -_prn[i];
+                }
+        }
+}
+
+
+void beidou_b1c_pilot_sinboc_61_gen_int(gsl::span<int> _dest, gsl::span<const int> _prn)
+{
+    const uint32_t _length_in = BEIDOU_B1Cp_CODE_LENGTH_CHIPS;
+    auto _period = static_cast<uint32_t>(_dest.size() / _length_in);
+
+    for (uint32_t i = 0; i < _length_in; i++)
+        {
+            for (uint32_t j = 0; j < _period; j += 2)
+                {
+                    _dest[i * _period + j] = _prn[i];
+                }
+            for (uint32_t j = 1; j < _period; j += 2)
+                {
+                    _dest[i * _period + j] = -_prn[i];
+                }
+        }
+}
+
+
+//! Generates float version of sine BOC(1,1) modulated Data Code 
+void beidou_b1cd_gen_float_11(gsl::span<float> _dest, gsl::span<int> _prn)
+{
+	const uint32_t _codeLength = 12 * BEIDOU_B1Cd_CODE_LENGTH_CHIPS;
+	
+	const float alpha = (1.0 / 2.0);
+	
+	int32_t sinboc_11[12 * 10230] = {0}; //  _codeLength not accepted by Clang
+	
+	gsl::span<int32_t> sinboc_11_(sinboc_11, _codeLength);
+	
+	beidou_b1c_data_sinboc_11_gen_int(sinboc_11_, _prn); //generate sinboc(1,1) 12 samples per chip
+	
+	for (uint32_t i = 0; i < _codeLength; i++)
+        {
+        	_dest[i] = alpha * static_cast<float>(sinboc_11[i]);
+                               
+	}
+}
+
+
+void beidou_b1cd_code_gen_float_sampled_boc_11(gsl::span<float> _dest, uint32_t _prn)
+{
+    // This function is based on the GNU software GPS for MATLAB in Kay Borre's book
+    auto _codeLength = static_cast<uint32_t>(BEIDOU_B1Cd_CODE_LENGTH_CHIPS);
+    auto* b1c_data_primary_code_chips = static_cast<int32_t*>(volk_gnsssdr_malloc(static_cast<uint32_t>(BEIDOU_B1Cd_CODE_LENGTH_CHIPS) * sizeof(int32_t), volk_gnsssdr_get_alignment()));
+    const int32_t _samplesPerChip =  12;
+
+    make_b1cd(gsl::span<int32_t>(b1c_data_primary_code_chips, static_cast<uint32_t>(BEIDOU_B1Cd_CODE_LENGTH_CHIPS)), _prn); //generate Beidou B1C code, 1 sample per chip
+
+    _codeLength = _samplesPerChip * BEIDOU_B1Cd_CODE_LENGTH_CHIPS;
+    std::unique_ptr<float> _signal_B1C_data{new float[_codeLength]};
+    gsl::span<float> _signal_B1C_data_span(_signal_B1C_data, _codeLength);
+
+    beidou_b1cd_gen_float_11(_signal_B1C_data_span, gsl::span<int>(b1c_data_primary_code_chips, static_cast<uint32_t>(BEIDOU_B1Cd_CODE_LENGTH_CHIPS)));  // generate cboc 12 samples per chip
+    
+    volk_gnsssdr_free(b1c_data_primary_code_chips);
```

Comment:
Avoid this memory allocation scheme, and keep same consistency as per the allocations you did before. See how B2a does this, where you define a code length and then allocate an array  based on that size